pipeline/exec/sharedcache.go

```diff
@@ -0,0 +1,170 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	pbbstream "github.com/streamingfast/bstream/pb/sf/bstream/v1"
+	"github.com/streamingfast/substreams/metrics"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+	"github.com/streamingfast/substreams/wasm"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+var GlobalSharedCache *SharedCache
+
+type SharedCache struct {
+	sync.Mutex
+	headBlock *atomic.Uint64
+	size      uint64
+	// module_hash -> blockHash -> callEntry
+	callEntries map[string]map[string]*callEntry
+}
+
+const maxUint64 uint64 = 1<<64 - 1
+
+func NewSharedCache(size uint64) *SharedCache {
+	return &SharedCache{
+		headBlock:   atomic.NewUint64(maxUint64),
+		size:        size,
+		callEntries: make(map[string]map[string]*callEntry),
+	}
+}
+
+func (s *SharedCache) ProcessBlock(blk *pbbstream.Block, _ interface{}) error {
+	if tracer.Enabled() {
+		zlog.Debug("shared_cache process_block", zap.Uint64("block_num", blk.Number))
+	}
+	s.headBlock.Store(blk.Number)
+	s.cleanup(blk.Number)
+	return nil
+}
+
+func (s *SharedCache) cleanup(head uint64) {
+	if head < s.size {
+		return
+	}
+	// cleanup
+	lowest := head - s.size
+	s.Lock()
+	for clockID, hashesInClock := range s.callEntries {
+		for entryID, entry := range hashesInClock {
+			if entry.clock.Number < lowest {
+				delete(hashesInClock, entryID)
+				if tracer.Enabled() {
+					zlog.Debug("deleting entry", zap.Uint64("clock_num", entry.clock.Number))
+				}
+			}
+		}
+		if len(hashesInClock) == 0 {
+			delete(s.callEntries, clockID)
+			if tracer.Enabled() {
+				zlog.Debug("deleting clockID", zap.String("clock_id", clockID))
+			}
+		}
+
+	}
+	s.Unlock()
+}
+
+type callEntry struct {
+	sync.RWMutex
+	clock          *pbsubstreams.Clock
+	moduleName     string
+	entrypoint     string
+	logs           []string
+	logsByteCount  uint64
+	executionStack []string
+	returnValue    []byte
+	executed       bool
+
+	err      error
+	panicErr *wasm.PanicError
+}
+
+func applyResult(res *callEntry, call *wasm.Call) error {
+	if call.Clock.Id != res.clock.Id ||
+		call.Entrypoint != res.entrypoint ||
+		call.ModuleName != res.moduleName {
+		panic(fmt.Sprintf("invalid shared cache data on block %s (%s) for module %s (%s)", call.Clock, res.clock, call.ModuleName, res.moduleName))
```

Comment:
I understand this should not happen, but since we already have a returned error here, wouldn't it better to not crash everything, what do you think?